src/jupyterlabcontroller/models/v1/prepuller.py

```diff
@@ -1,152 +1,104 @@
         example="Latest Daily (Daily 2077_10_23)",
         title="Human-readable version of image tag",
     )
-    digest: str = Field(
-        ...,
-        name="digest",
+    digest: Optional[str] = Field(
+        None,
         example=(
             "sha256:e693782192ecef4f7846ad2b21"
             "b1574682e700747f94c5a256b5731331a2eec2"
         ),
-        title="unique digest of image contents",
+        title="Digest of image",
     )
 
 
-class Image(PartialImage):
-    tags: TagToNameMap = Field(
-        ...,
-        name="tags",
-        title="Map between tag and its display name",
-    )
-    size: Optional[int] = Field(
-        None,
-        name="size",
-        example=8675309,
-        title="Size in bytes of image.  None if image size is unknown",
+class PrepulledImage(Image):
+    """Used to display available images."""
+
+    aliases: list[str] = Field(
+        [], title="Other aliases", example=["recommended", "latest_weekly"]
     )
     prepulled: bool = Field(
         False,
         name="prepulled",
-        example=False,
+        example=True,
         title="Whether image is prepulled to all eligible nodes",
     )
 
-    @property
-    def references(self) -> List[str]:
-        r = [f"{self.path}@{self.digest}"]
-        for tag in self.tags:
-            r.append(f"{self.path}:{tag}")
-        return r
-
-
-"""GET /nublado/spawner/v1/images"""
-
-
-class SpawnerImages(CamelCaseModel):
-    recommended: Optional[Image] = None
-    latest_weekly: Optional[Image] = None
-    latest_daily: Optional[Image] = None
-    latest_release: Optional[Image] = None
-    all: List[Image] = Field(default_factory=list)
-
-    class Config:
-        alias_generator = dashify
-        allow_population_by_field_name = True
-
-
-"""GET /nublado/spawner/v1/prepulls"""
 
+class NodeImage(Image):
+    size: Optional[int] = Field(
+        None, example=8675309, title="Size in bytes of image if known"
+    )
+    nodes: list[str] = Field([], title="Nodes on which image is cached")
+    missing: Optional[list[str]] = Field(
+        None, title="Nodes not caching the image"
+    )
 
-# "config" section
 
-# This comes from PrepullerConfiguration
+class SpawnerImages(BaseModel):
+    recommended: PrepulledImage = Field(..., title="Recommended image")
+    latest_weekly: Optional[PrepulledImage] = Field(
+        None, title="Latest weekly"
+    )
+    latest_daily: Optional[PrepulledImage] = Field(None, title="Latest daily")
+    latest_release: Optional[PrepulledImage] = Field(
+        None, title="Latest release"
+    )
+    all: list[PrepulledImage] = Field(..., title="All available images")
 
 
-# "images" section
+class PrepullerImageStatus(BaseModel):
+    prepulled: list[NodeImage] = Field([], title="Successfully cached images")
+    pending: list[NodeImage] = Field(
+        [],
+        title="Images not yet cached on all eligible nodes",
+    )
 
 
-class Node(CamelCaseModel):
+class Node(BaseModel):
```

Comment:
Don't forget to update sqr-066 to reflect any changes made here to the externally-visible models.

Reply:
Yeah, good call, draft PR here: https://github.com/lsst-sqre/sqr-066/pull/3

This currently has some stuff in it from upcoming PRs.